### PR TITLE
SVCDBREP-54: Allow dynamo and CW credentials override

### PR DESCRIPTION
**Pull request by mikegirkin**

### CODE REVIEW

File: src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala [hunk 1]
```diff
@@ -30,10 +30,11 @@ object KCLConfiguration {
...
     */
-  def apply( applicationName: String
-           , streamName: String
-           , credentialsProvider: AWSCredentialsProvider = new DefaultAWSCredentialsProviderChain()
-           ): KinesisClientLibConfiguration = {
+  def apply(
+    applicationName: String,
+    streamName: String,
+    credentialsProvider: AWSCredentialsProvider = new DefaultAWSCredentialsProviderChain()
+  ): KinesisClientLibConfiguration = {
```
Comment (gheine) [inline src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala]: Revert formatting changes

File: src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLWorkerRunner.scala [hunk 2]
```diff
@@ -23,11 +26,12 @@ import scala.util.{Failure, Success, Try}
...
-, checkpointInterval: FiniteDuration = 5 minutes
-, numRetries: Int = 3
-, initialize: (String) => Unit = (_) => ()
-, shutdown: (String, IRecordProcessorCheckpointer, ShutdownReason) => Unit = (_,_,_) => ()
+  config: KinesisClientLibConfiguration,
+  checkpointInterval: FiniteDuration = 5 minutes,
+  numRetries: Int = 3,
+  initialize: (String) => Unit = (_) => (),
+  shutdown: (String, IRecordProcessorCheckpointer, ShutdownReason) => Unit = (_,_,_) => (),
+  metricsFactory: Option[IMetricsFactory] = None
```
Comment (gheine) [inline src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLWorkerRunner.scala]: Revert formatting changes

File: src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLWorkerRunner.scala [hunk 3]
```diff
@@ -68,9 +82,9 @@ case class KCLWorkerRunner (
    * @param processRecords     (ShardId, Records, Checkpointer) => Unit : Kinesis record handler
    * @param evReader           evidence that A has implementation of KinesisRecordReader implicitly available in scope
    */
-  def runBatchProcessor[A]( processRecords: (String, Seq[A], IRecordProcessorCheckpointer) => Unit
-                         )( implicit evReader: KinesisRecordReader[A]
-                          ): Unit = {
+  def runBatchProcessor[A]
+    (processRecords: (String, Seq[A], IRecordProcessorCheckpointer) => Unit)
+    (implicit evReader: KinesisRecordReader[A]): Unit = {
```
Comment (gheine) [inline src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLWorkerRunner.scala]: Revert formatting changes

File: src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala [hunk 4]
```diff
@@ -44,4 +45,23 @@ object KCLConfiguration {
     , s"${HostName}:${UUID.randomUUID()}"
     )
   }
+
+  def apply(
+    applicationName: String,
+    streamName: String,
+    kinesisCredentialsProvider: AWSCredentialsProvider,
+    dynamoCredentialsProvider: AWSCredentialsProvider,
+    cloudWatchCredentialsProvider: AWSCredentialsProvider
```
Comment (gheine) [inline src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala]: Why not add the 2 new credentials providers to the existing apply method with the same default value "new DefaultAWSCredentialsProviderChain()"?

File: src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLWorkerRunner.scala [hunk 5]
```diff
@@ -61,6 +65,16 @@ case class KCLWorkerRunner (
...
 
+  def withMetricsFactory(factory: IMetricsFactory): KCLWorkerRunner = {
+    this.copy(metricsFactory = Some(factory))
+  }
+
+
+  protected def createWorker(recordProcessorFactory: IRecordProcessorFactory) = {
+    metricsFactory.fold(
+      new Worker(recordProcessorFactory, config))(
+      metricsFactory => new Worker(recordProcessorFactory, config, metricsFactory))
```
Comment (gheine) [inline src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLWorkerRunner.scala]: Use a different name for the lambda function argument (e.g. "mf") so there's no ambiguity which "metricsFactory" is passed into the new Worker.

**Comment by vdumitrescu** [comment]
LGTM.

File: src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLWorkerRunner.scala [hunk 6]
```diff
@@ -61,6 +64,16 @@ case class KCLWorkerRunner (
     this.copy(shutdown = sd)
   }
 
+  def withMetricsFactory(factory: IMetricsFactory): KCLWorkerRunner = {
+    this.copy(metricsFactory = Some(factory))
+  }
+
+
+  protected def createWorker(recordProcessorFactory: IRecordProcessorFactory) = {
```
Comment (gheine) [inline src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLWorkerRunner.scala]: There seems to be no need to expose this as part of the (protected) API. Make this private or an inner function in runBatchProcessor or simply inline it in runBatchProcessor.

File: src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLWorkerRunner.scala [hunk 7]
```diff
@@ -89,7 +102,7 @@ case class KCLWorkerRunner (
         errs.map(_.failed.get).foreach { e => error(e.getMessage, e) }
       }
 
-      (new Worker(recordProcessorFactory, config)).run()
+      createWorker(recordProcessorFactory).run()
```
Comment (gheine) [inline src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLWorkerRunner.scala]: val worker = metricsFactory.fold(new Worker(recordProcessorFactory, config))(mf => new Worker(recordProcessorFactory, config, mf))
worker.run()

**Comment by gheine** [comment]
:+1:

**Pull request merged**